DllImportGenerator/DllImportGenerator/Analyzers/ManualTypeMarshallingAnalyzer.cs

```diff
@@ -343,21 +366,29 @@ private void AnalyzeNativeMarshalerType(SymbolAnalysisContext context, ITypeSymb
                 // Validate that the native type has at least one marshalling method (either managed to native or native to managed)
                 if (!hasConstructor && !hasStackallocConstructor && !hasToManaged)
                 {
-                    context.ReportDiagnostic(Diagnostic.Create(NativeTypeMustHaveRequiredShapeRule, GetSyntaxReferenceForDiagnostic(marshalerType).GetSyntax().GetLocation(), marshalerType.ToDisplayString(), type.ToDisplayString()));
+                    context.ReportDiagnostic(Diagnostic.Create(NativeTypeMustHaveRequiredShapeRule,
+                        marshalerType.Locations[0],
+                        marshalerType.Locations.Skip(1),
+                        marshalerType.ToDisplayString(), type.ToDisplayString()));
                 }
 
                 // Validate that this type can support marshalling when stackalloc is not usable.
                 if (validateAllScenarioSupport && hasStackallocConstructor && !hasConstructor)
                 {
-                    context.ReportDiagnostic(Diagnostic.Create(StackallocMarshallingShouldSupportAllocatingMarshallingFallbackRule, GetSyntaxReferenceForDiagnostic(marshalerType).GetSyntax().GetLocation(), marshalerType.ToDisplayString()));
+                    context.ReportDiagnostic(Diagnostic.Create(StackallocMarshallingShouldSupportAllocatingMarshallingFallbackRule,
+                        marshalerType.Locations[0],
+                        marshalerType.Locations.Skip(1),
+                        marshalerType.ToDisplayString()));
                 }
 
                 IPropertySymbol? valueProperty = ManualTypeMarshallingHelper.FindValueProperty(nativeType);
                 bool valuePropertyIsRefReturn = valueProperty is { ReturnsByRef : true } or { ReturnsByRefReadonly: true };
 
                 if (valuePropertyIsRefReturn)
```

Comment:
Let's move this predicate into the check for `valueProperty is not null` below. This is a great example of why I dislike the `?` types. Null checks become implicit in some cases but not in others and it breeds redundant checks that aren't obvious or clear when they occur and thus muddles clear logical flow.